File: aws-globalaccelerator-accelerator/src/main/java/software/amazon/globalaccelerator/accelerator/ReadHandler.java
package software.amazon.globalaccelerator.accelerator;

import com.amazonaws.services.globalaccelerator.model.Accelerator;
import lombok.val;
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
import software.amazon.cloudformation.proxy.HandlerErrorCode;
import software.amazon.cloudformation.proxy.Logger;
import software.amazon.cloudformation.proxy.ProgressEvent;
import software.amazon.cloudformation.proxy.ResourceHandlerRequest;

import java.util.stream.Collectors;

public class ReadHandler extends BaseHandler<CallbackContext> {

    @Override
    public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
        final AmazonWebServicesClientProxy proxy,
        final ResourceHandlerRequest<ResourceModel> request,
        final CallbackContext callbackContext,
        final Logger logger) {

        val agaClient = AcceleratorClientBuilder.getClient();
        final ResourceModel model = request.getDesiredResourceState();
        logger.log(String.format("Read request for accelerator: [%s]", request));

        val accelerator = HandlerCommons.getAccelerator(model.getAcceleratorArn(), proxy, agaClient, logger);
        val acceleratorResourceModel = convertAcceleratorToResourceModel(accelerator);

        if (acceleratorResourceModel == null){
            logger.log(String.format("Accelerator with ARN [%s] not found", model.getAcceleratorArn()));
            return ProgressEvent.defaultFailureHandler(new Exception("Listener not found."), HandlerErrorCode.NotFound);
        } else {
            return ProgressEvent.defaultSuccessHandler(acceleratorResourceModel);
        }
    }
    private ResourceModel convertAcceleratorToResourceModel(Accelerator accelerator) {
        ResourceModel newModel = null;
        if (accelerator != null) {
            newModel = new ResourceModel();
            newModel.setAcceleratorArn(accelerator.getAcceleratorArn());
            newModel.setName(accelerator.getName());
            newModel.setEnabled(accelerator.getEnabled());
            newModel.setIpAddressType(accelerator.getIpAddressType());
            newModel.setIpAddresses(accelerator.getIpSets() == null ? null :
                            accelerator.getIpSets().stream()
                            .flatMap(x -> x.getIpAddresses().stream())
                            .collect(Collectors.toList()));
        }
        return newModel;
    }
}

File: aws-globalaccelerator-accelerator/src/main/kotlin/software/amazon/globalaccelerator/accelerator/ReadHandler.kt
package software.amazon.globalaccelerator.accelerator

import com.amazonaws.services.globalaccelerator.model.Accelerator
import com.amazonaws.services.globalaccelerator.model.Tag
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.HandlerErrorCode
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.ProgressEvent
import software.amazon.cloudformation.proxy.ResourceHandlerRequest
import software.amazon.globalaccelerator.accelerator.AcceleratorClientBuilder.client
import software.amazon.globalaccelerator.accelerator.HandlerCommons.getAccelerator
import software.amazon.globalaccelerator.accelerator.HandlerCommons.getTags
import software.amazon.globalaccelerator.accelerator.HandlerCommons.toResourceModel

/**
 * Read handler implementation for accelerator resource.
 */
class ReadHandler : BaseHandler<CallbackContext?>() {
    override fun handleRequest(proxy: AmazonWebServicesClientProxy,
                               request: ResourceHandlerRequest<ResourceModel>,
                               callbackContext: CallbackContext?,
                               logger: Logger): ProgressEvent<ResourceModel, CallbackContext?> {
        logger.debug("Read Accelerator Request: $request")
        val agaClient = client
        val model = request.desiredResourceState
        val accelerator = getAccelerator(model.acceleratorArn, proxy, agaClient, logger)
        val tags = getTags(model.acceleratorArn, proxy, agaClient, logger)
        val acceleratorResourceModel = convertAcceleratorToResourceModel(accelerator, tags)
        return if (acceleratorResourceModel == null) {
            logger.error("Accelerator with arn: [${model.acceleratorArn}] not found.")
            ProgressEvent.defaultFailureHandler(Exception("Accelerator not found."), HandlerErrorCode.NotFound)
        } else {
            ProgressEvent.defaultSuccessHandler(acceleratorResourceModel)
        }
    }

    private fun convertAcceleratorToResourceModel(accelerator: Accelerator?, tags: List<Tag>): ResourceModel? {
        return accelerator?.let { toResourceModel(it, tags) }
    }
}
